lib/ocpp/v201/charge_point.cpp

```diff
@@ -3207,6 +3232,96 @@ void ChargePoint::handle_set_charging_profile_req(Call<SetChargingProfileRequest
+}
+
+void ChargePoint::handle_get_charging_profiles_req(Call<GetChargingProfilesRequest> call) {
+    EVLOG_debug << "Received GetChargingProfilesRequest: " << call.msg << "\nwith messageId: " << call.uniqueId;
+    const auto msg = call.msg;
+    GetChargingProfilesResponse response;
+
+    const auto profiles_to_report = this->smart_charging_handler->get_reported_profiles(msg);
+
+    response.status =
+        profiles_to_report.empty() ? GetChargingProfileStatusEnum::NoProfiles : GetChargingProfileStatusEnum::Accepted;
+
+    ocpp::CallResult<GetChargingProfilesResponse> call_result(response, call.uniqueId);
+    this->send<GetChargingProfilesResponse>(call_result);
+
+    if (response.status == GetChargingProfileStatusEnum::NoProfiles) {
+        return;
+    }
+
+    // There are profiles to report.
+    // Prepare ReportChargingProfileRequest(s). The message defines the properties evseId and chargingLimitSource as
+    // required, so we can not report all profiles in a single ReportChargingProfilesRequest. We need to prepare a
+    // single ReportChargingProfilesRequest for each combination of evseId and chargingLimitSource
+    std::set<int32_t> evse_ids;                // will contain all evse_ids of the profiles
+    std::set<ChargingLimitSourceEnum> sources; // will contain all sources of the profiles
+
+    // fill evse_ids and sources sets
+    std::transform(profiles_to_report.begin(), profiles_to_report.end(), std::inserter(evse_ids, evse_ids.end()),
+                   [](const ReportedChargingProfile& profile) { return profile.evse_id; });
+    std::transform(profiles_to_report.begin(), profiles_to_report.end(), std::inserter(sources, sources.end()),
+                   [](const ReportedChargingProfile& profile) { return profile.source; });
+
+    std::vector<ReportChargingProfilesRequest> requests_to_send;
+
+    for (const auto evse_id : evse_ids) {
+        for (const auto source : sources) {
+            std::vector<ChargingProfile> original_profiles;
+            std::for_each(profiles_to_report.begin(), profiles_to_report.end(),
+                          [evse_id, source, &original_profiles](ReportedChargingProfile reported_profile) {
+                              if (reported_profile.evse_id == evse_id and reported_profile.source == source) {
+                                  original_profiles.push_back(reported_profile.profile);
+                              };
+                          });
+            if (not original_profiles.empty()) {
+                // prepare a ReportChargingProfilesRequest
+                ReportChargingProfilesRequest req;
+                req.requestId = msg.requestId; // K09.FR.01
+                req.evseId = evse_id;
+                req.chargingLimitSource = source;
+                req.chargingProfile = original_profiles;
+                requests_to_send.push_back(req);
+            }
+        }
+    }
+
+    // requests_to_send are ready, send them and define tbc property
```

Comment:
I think you can set tbc default to true and then do `requests_to_send.back().tbc = false`. Then its just iterating all to send.